lib/membrane_opus/parser.ex

```diff
@@ -62,149 +70,92 @@ defmodule Membrane.Opus.Parser do
-    input_delimitation =
-      Map.get(ctx.pads.input, :caps) || %{} |> Map.get(:self_delimiting?, state.input_delimitted?)
-
-    {configuration_number, stereo_flag, frame_packing, data_without_toc} = parse_toc_byte(data)
-    {_mode, _bandwidth, frame_duration} = parse_configuration(configuration_number)
-    channels = parse_channels(stereo_flag)
-
-    {frame_lengths, header_size} =
-      FrameLengths.parse_frame_lengths(frame_packing, data_without_toc, input_delimitation)
-
-    {parsed_data, self_delimiting?} =
-      parse_data(data, frame_lengths, header_size, state, input_delimitation)
+  def handle_caps(:input, caps, _ctx, state) when not state.force_reading_input_as_delimited? do
+    {:ok, %{state | input_delimited?: caps.self_delimiting?}}
+  end
 
-    caps = %Opus{
-      channels: channels,
-      self_delimiting?: self_delimiting?
-    }
+  @impl true
+  def handle_process(:input, %Buffer{payload: data}, _ctx, state) do
+    {delimitation_handler, self_delimiting?} =
+      Delimitation.get_handler(state.delimitation, state.input_delimited?)
+
+    {buffer, packets, channels} =
+      maybe_parse(state.buffer <> data, state.input_delimited?, delimitation_handler)
+
+    packet_actions =
+      if length(packets) > 0 do
+        [
+          caps:
+            {:output,
+             %Opus{
+               channels: channels,
+               self_delimiting?: self_delimiting?
+             }},
+          buffer: {:output, packets}
+        ]
+      else
+        []
+      end
+
+    redemand_actions =
+      if byte_size(buffer) > 0 do
+        [redemand: :output]
+      else
+        []
+      end
+
+    actions = packet_actions ++ redemand_actions
+
+    # it's an invariant of this filter that we send packets and/or need to
+    # redemand
+    if length(actions) > 0 do
+      {{:ok, actions}, %{state | buffer: buffer}}
+    else
+      {:error, "Invalid input"}
+    end
+  end
 
-    buffer = %Buffer{
-      payload: parsed_data,
-      metadata: %{
-        duration: elapsed_time(frame_lengths, frame_duration)
+  @spec maybe_parse(
+          data :: binary,
+          input_delimited? :: boolean,
+          handler :: Delimitation.handler_t(),
+          packets :: [Buffer.t()],
+          channels :: 0..2
+        ) :: {remaining_buffer :: binary, packets :: [Buffer.t()], channels :: 0..2}
+  defp maybe_parse(data, input_delimited?, handler, packets \\ [], channels \\ 0) do
+    with {configuration_number, stereo_flag, frame_packing} <- Util.parse_toc_byte(data),
+         channels <- max(channels, Util.parse_channels(stereo_flag)),
+         {_mode, _bandwidth, frame_duration} <- Util.parse_configuration(configuration_number),
+         {header_size, frame_lengths, padding_size} <-
+           FrameLengths.parse(frame_packing, data, input_delimited?),
+         expected_packet_size <- header_size + Enum.sum(frame_lengths) + padding_size,
+         <<raw_packet::binary-size(expected_packet_size), rest::binary>> <- data do
```

Comment:
I get a warning here
> variable "raw_packet" is unused

This is actually quite strange that everything works anyway